src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,170 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+  //     X                  [X, 3]       // Index 3 points past node input array which means that there are no more inputs
+  //                                     // to visit. Once X is on top of stack again, we are done with 'X' and pop it.
+  //
+  // If an input was already collected before (i.e. part of start->target), then the current node is part of some kind
+  // of diamond with it:
+  //
+  //        C3       X
+  //       /  \     /
+  //     C2   current
+  //      \  /
+  //       C1
+  //
+  // Cx means collected. Since C3 is already collected (and thus already visited), we add 'current' to the collected list
+  // since it must also be on a path from start->target. We continue the DFS with X which could potentially also be on a
+  // start->target path but that is not known yet.
+  //
+  // This method returns true if there is an unvisited input and return false otherwise if all inputs have been visited.
+  bool push_next_unvisited_input() {
+    Node* current = _stack.node();
+    const uint next_unvisited_input_index = _stack.index();
+    for (uint i = next_unvisited_input_index; i < current->req(); i++) {
+      Node* input = current->in(i);
+      if (_node_filter(input)) {
+        if (!_visited.test(input->_idx)) { // Avoid double visits which could take a long time to process.
+          // Visit current->in(i) next in DFS walk. Once 'current' is again on top of stack, we need to visit in(i+1).
+          push_input_and_update_current_index(input, i + 1);
+          return true;
+        } else if (_collected_nodes.member(input)) {
+          // Diamond case, see description above.
+          // Input node part of start->target? Then current node (i.e. a predecessor of input) is also on path. Collect it.
+          _collected_nodes.push(current);
+        }
+      }
+    }
+    return false;
+  }
+
+  // Update the index of the current node on top of the stack with the next unvisited input index and push 'input' to
+  // the stack which is visited next in the DFS order.
+  void push_input_and_update_current_index(Node* input, uint next_unvisited_input_index) {
+    _stack.set_index(next_unvisited_input_index);
+    push_unvisited_node(input);
+  }
+
+  // Push the next unvisited node in the DFS order with index 1 since this node needs to visit all its inputs.
+  void push_unvisited_node(Node* next_to_visit) {
+    _stack.push(next_to_visit, 1);
+  }
+
+  // If the current node on top of the stack is on a path from start->target(s), then also collect the predecessor node
+  // before popping the current node.
+  void pop_node_and_maybe_collect_predecessor() {
+    Node* current_node = _stack.node();
+    _stack.pop();
+    if (_stack.is_nonempty() && _collected_nodes.member(current_node)) {
+      // Current node was part of start->target? Then predecessor (i.e. newly on top of stack) is also on path. Collect it.
+      Node* predecessor = _stack.node();
+      _collected_nodes.push(predecessor);
+    }
+  }
+};
```

Comment:
I think this is now correct. But it is 100 lines to perform and explain this DFS with backtracking.

On the other hand doing 2 BFS would just be 20+ lines.

```
Unique_Node_List collected;

Unique_Node_List input_traversal;
input_traversal.push(start_node);
for (int i = 0; i < input_traversal.length(); i++) {
  Node* n = input_traversal.at(i);
  for (int j = 1; j < n->req(); j++) {
    Node* input = n->in(j);
    if (_is_target_node(input)) {
      collected.push(input); // mark as target, where we start backtracking.
    } else if(_filter(input)) {
      input_traversal.push(input); // continue BFS.
    }
  }
}
assert(!collected.is_empty(), "must find some targets");
for (int i = 0; i < collected.length(); i++) {
  Node* n = collected.at(i);
  for (output : n->fastout()) { // pseudocode
    if (input_traversal.contains(output)) {
      collected.push(output); // backtrack through nodes of input traversal
    }
  }
}
assert(collected.contains(start_node), "must find start node again");
```

Reply:
But this is a matter of taste. The data structures probably have roughly the same size. And also runtime is probably basically the same.

Reply:
I think you're right. The typical Template Assertion Predicate Expression only has few nodes. So, I totally agree with you that we should go with a much simpler double BFS version instead of a single DFS walk. And if we ever find the need to have a single DFS walk with backtracking where a single walk is expensive, we could still go back to this PR and revive this code.

Updated code.